fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Ordering.java

```diff
@@ -0,0 +1,515 @@
+/*
+ * Ordering.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.temp;
+
+import com.apple.foundationdb.record.metadata.expressions.KeyExpression;
+import com.apple.foundationdb.record.query.expressions.Comparisons.Comparison;
+import com.google.common.base.Verify;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSetMultimap;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.PeekingIterator;
+import com.google.common.collect.SetMultimap;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.BinaryOperator;
+
+/**
+ * This class captures an ordering property. An ordering is a list of parts that define the actual ordering of
+ * records together with a (multi-)set of equality-bound keys. The expressions that are bound by equality is
+ * disjoint from the set of expressions partaking in the list of orderings.
+ *
+ * <h2>
+ * Compliant Example
+ * </h2>
+ * <pre>
+ * {@code
+ *   ordering parts: (a, b, c); equality-bound: d = 5, e = $p*
+ * }
+ * </pre>
+ *
+ * <h2>
+ * Non-Compliant Example
+ * </h2>
+ * <pre>
+ * {@code
+ *   ordering parts: (a, b, c); equality-bound: c = 5
+ * }
+ * </pre>
+ *
+ * Note that such an ordering would always be:
+ * <pre>
+ * {@code
+ *   ordering parts: (a, b); equality-bound: c = 5
+ * }
+ *
+ * </pre>
+ *
+ * Instances of this class are used to communicate properties of plans as well as required properties.
+ *
+ */
+public class Ordering {
```

Comment:
Since we are introducing this new abstraction, perhaps this would be a good time to make the ordering parts _each_ include ascending / descending, rather than relying on a separate reversed flag to tag along next to all this.

Reply:
follow up issue #1318